storage/storagetest/storagetest.go

```diff
@@ -0,0 +1,277 @@
+	ctx := context.Background()
+
+	// Test retrieval methods on empty storage.
+	t.Run("empty", func(t *testing.T) {
+		const name = "empty"
+
+		storage := storageFn(t)
+
+		pemCert, pemKey, err := storage.RetrieveTokenPKI(ctx, name)
+		// Currently just checking that it returns an error when the resource is not found.
+		//
+		// TODO(lucas): Define an "resource not found" error type as part of the interface
+		// (to be returned by all implementations).
+		if err == nil {
+			t.Fatal("expected error")
+		}
+		if pemCert != nil {
+			t.Fatal("expected nil cert pem")
+		}
+		if pemKey != nil {
+			t.Fatal("expected nil key pem")
+		}
+
+		// Currently just checking that it returns an error when the resource is not found.
+		//
+		// TODO(lucas): Define an "resource not found" error type as part of the interface
+		// (to be returned by all implementations).
+		tokens, err := storage.RetrieveAuthTokens(ctx, name)
+		if err == nil {
+			t.Fatal("expected error")
+		}
+		if tokens != nil {
+			t.Fatal("expected nil tokens")
+		}
+
+		config, err := storage.RetrieveConfig(ctx, name)
+		checkErr(t, err)
+		emptyConfig := client.Config{}
+		if config == nil || *config != emptyConfig {
+			t.Fatalf("expected empty config: %+v", config)
+		}
+
+		// Profile assigner storing and retrieval.
+		profileUUID, modTime, err := storage.RetrieveAssignerProfile(ctx, name)
+		checkErr(t, err)
+		if profileUUID != "" {
+			t.Fatal("expected empty profileUUID")
+		}
+		if !modTime.IsZero() {
+			t.Fatal("expected zero modTime")
+		}
+
+		cursor, err := storage.RetrieveCursor(ctx, name)
+		checkErr(t, err)
+		if cursor != "" {
+			t.Fatal("expected empty cursor")
+		}
+	})
+
+	testWithName := func(t *testing.T, name string, storage storage.AllStorage) {
+
+		// PKI storing and retrieval.
+		pemCert, pemKey, err := storage.RetrieveTokenPKI(ctx, name)
+		// Currently just checking that it returns an error when the resource is not found.
+		//
+		// TODO(lucas): Define an "resource not found" error type as part of the interface
+		// (to be returned by all implementations).
+		if err == nil {
+			t.Fatal("expected error")
+		}
+		if pemCert != nil {
+			t.Fatal("expected nil cert pem")
+		}
+		if pemKey != nil {
+			t.Fatal("expected nil key pem")
+		}
+		pemCert, pemKey = generatePKI(t, "basicdn", 1)
+		err = storage.StoreTokenPKI(ctx, name, pemCert, pemKey)
+		checkErr(t, err)
+		pemCert2, pemKey2, err := storage.RetrieveTokenPKI(ctx, name)
+		checkErr(t, err)
+		if !bytes.Equal(pemCert, pemCert2) {
+			t.Fatalf("pem cert mismatch: %s vs. %s", pemCert, pemCert2)
+		}
+		if !bytes.Equal(pemKey, pemKey2) {
+			t.Fatalf("pem key mismatch: %s vs. %s", pemKey, pemKey2)
+		}
+
+		// Token storing and retrieval.
+		tokens, err := storage.RetrieveAuthTokens(ctx, name)
+		// Currently just checking that it returns an error when the resource is not found.
+		//
+		// TODO(lucas): Define an "resource not found" error type as part of the interface
```

Comment:
Good idea. Waiting on feedback from @jessepeterson to do that?

Reply:
Oh I added it but wanted to check first if we all agree it's a good idea to define it and return it in all methods when the resource (config, tokens, pki) is not found.